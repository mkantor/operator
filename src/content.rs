extern crate handlebars;
use crate::lib::*;
use handlebars::{Context, Handlebars, RenderContext, Renderable};
use serde::Serialize;
use std::io;
use std::path::{Path, PathBuf};
use thiserror::Error;

#[derive(Error, Debug)]
#[error(
  "Failed to parse template{} from content directory path '{}'.",
  .source.template_name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or_default(),
  .content_directory_path.to_string_lossy()
)]
pub struct RegisteredTemplateParseError {
    source: handlebars::TemplateError,
    content_directory_path: PathBuf,
}

#[derive(Error, Debug)]
#[error(
  "Failed to parse template{}.",
  .source.template_name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or_default(),
)]
pub struct UnregisteredTemplateParseError {
    source: handlebars::TemplateError,
}

#[derive(Error, Debug)]
pub enum ContentLoadingError {
    #[error(transparent)]
    TemplateParseError(#[from] RegisteredTemplateParseError),

    #[error(
    "Input/output error when loading{} from content directory path '{}'.",
    .name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or(String::from(" content")),
    .content_directory_path.to_string_lossy()
  )]
    IOError {
        source: io::Error,
        content_directory_path: PathBuf,
        name: Option<String>,
    },
}

#[derive(Error, Debug)]
#[error(
  "Rendering failed for template{}.",
  .source.template_name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or_default(),
)]
pub struct TemplateRenderError {
    #[from]
    source: handlebars::RenderError,
}

pub struct ContentEngine<'a> {
    template_registry: Handlebars<'a>,
}

impl<'a> ContentEngine<'a> {
    pub fn from_content_directory(
        content_directory_path: &'a Path,
    ) -> Result<Self, ContentLoadingError> {
        let template_registry = {
            let mut template_registry = Handlebars::new();
            template_registry
                .register_templates_directory(".hbs", content_directory_path)
                .map_err(|template_render_error| match template_render_error {
                    handlebars::TemplateFileError::TemplateError(source) => {
                        ContentLoadingError::TemplateParseError(RegisteredTemplateParseError {
                            source,
                            content_directory_path: PathBuf::from(content_directory_path),
                        })
                    }
                    handlebars::TemplateFileError::IOError(source, original_name) => {
                        // Handlebars-rust will use an empty string when the error does not
                        // correspond to a specific path.
                        let name = if original_name.is_empty() {
                            None
                        } else {
                            Some(original_name)
                        };
                        ContentLoadingError::IOError {
                            source,
                            content_directory_path: PathBuf::from(content_directory_path),
                            name,
                        }
                    }
                })?;
            template_registry
        };

        Ok(ContentEngine { template_registry })
    }

    pub fn new_content(
        &self,
        handlebars_source: &str,
    ) -> Result<ContentItem, UnregisteredTemplateParseError> {
        let template = handlebars::Template::compile2(handlebars_source, true)
            .map_err(|source| UnregisteredTemplateParseError { source })?;
        Ok(ContentItem::UnregisteredTemplate {
            template_registry: &self.template_registry,
            template,
        })
    }
}

#[derive(Serialize)]
struct GluonRenderData {
    version: GluonVersion,
}

#[derive(Serialize)]
struct RenderData {
    gluon: GluonRenderData,
}

pub enum ContentItem<'a> {
    /// A named template that exists in the registry. This variant must only be
    /// constructed with names that have already been validated against the
    /// registry.
    RegisteredTemplate {
        template_registry: &'a Handlebars<'a>,
        name_in_registry: &'a str,
    },

    /// An anonymous template for on-the-fly rendering.
    UnregisteredTemplate {
        template_registry: &'a Handlebars<'a>,
        template: handlebars::Template,
    },
}

impl Render for ContentItem<'_> {
    type RenderArgs = GluonVersion;
    type Error = TemplateRenderError;

    fn render(&self, gluon_version: GluonVersion) -> Result<String, Self::Error> {
        let render_data = RenderData {
            gluon: GluonRenderData {
                version: gluon_version,
            },
        };

        let rendered_content = match &self {
            ContentItem::RegisteredTemplate {
                template_registry,
                name_in_registry,
            } => template_registry.render(name_in_registry, &render_data)?,
            ContentItem::UnregisteredTemplate {
                template_registry,
                template,
            } => {
                let context = Context::wraps(render_data)?;
                let mut render_context = RenderContext::new(None);
                template.renders(template_registry, &context, &mut render_context)?
            }
        };

        Ok(rendered_content)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_lib::*;

    #[test]
    fn content_engine_can_be_created_from_valid_content_directory() {
        for &path in &CONTENT_DIRECTORY_PATHS_WITH_VALID_CONTENTS {
            assert!(
                ContentEngine::from_content_directory(Path::new(path)).is_ok(),
                "Content engine could not be created from {}",
                path
            );
        }
    }

    #[test]
    fn content_engine_cannot_be_created_from_invalid_content_directory() {
        for &path in &CONTENT_DIRECTORY_PATHS_WITH_INVALID_CONTENTS {
            assert!(
                ContentEngine::from_content_directory(Path::new(path)).is_err(),
                "Content engine was successfully created from {}, but this should have failed",
                path
            );
        }
    }

    #[test]
    fn new_templates_can_be_rendered() {
        let engine = ContentEngine::from_content_directory(
            arbitrary_content_directory_path_with_valid_content(),
        )
        .expect("Content engine could not be created");

        for &(template, expected_output) in &VALID_TEMPLATES {
            let new_content = engine
                .new_content(template)
                .expect("Template could not be parsed");
            let rendered = new_content
                .render(GluonVersion("0.0.0"))
                .expect(&format!("Template rendering failed for `{}`", template,));
            assert_eq!(
                rendered,
                expected_output,
                "Template rendering for `{}` did not produce the expected output (\"{}\"), instead got \"{}\"",
                template,
                expected_output,
                rendered,
            );
        }
    }

    #[test]
    fn new_content_fails_for_invalid_templates() {
        let engine = ContentEngine::from_content_directory(
            arbitrary_content_directory_path_with_valid_content(),
        )
        .expect("Content engine could not be created");

        for &template in &INVALID_TEMPLATES {
            let result = engine.new_content(template);

            assert!(
                result.is_err(),
                "Content was successfully created for invalid template `{}`, but it should have failed",
                template,
            );
        }
    }

    #[test]
    fn new_templates_can_reference_partials_from_content_directory() {
        let content_directory_path = example_path("valid/partials");
        let engine = ContentEngine::from_content_directory(&content_directory_path)
            .expect("Content engine could not be created");

        let template = "this is partial: {{> ab}}";
        let expected_output = "this is partial: a\nb\n\n";

        let new_content = engine
            .new_content(template)
            .expect("Template could not be parsed");
        let rendered = new_content
            .render(GluonVersion("0.0.0"))
            .expect(&format!("Template rendering failed for `{}`", template,));
        assert_eq!(
            rendered,
            expected_output,
            "Template rendering for `{}` did not produce the expected output (\"{}\"), instead got \"{}\"",
            template,
            expected_output,
            rendered,
        );
    }
}
