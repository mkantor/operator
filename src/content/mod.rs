mod body;
mod content_directory;
mod content_engine;
mod content_index;
mod content_item;
mod content_registry;
mod handlebars_helpers;
mod mime;
mod route;
mod test_lib;

use crate::bug_message;
use bytes::Bytes;
use content_item::RenderingFailedError;
use futures::Stream;
use serde::Serialize;
use std::io;
use thiserror::Error;

pub use self::mime::{MediaRange, MediaType};
pub use content_directory::ContentDirectory;
pub use content_engine::{
    ContentEngine, ContentLoadingError, FilesystemBasedContentEngine, TemplateError,
};
pub use content_index::ContentIndex;
pub use content_item::UnregisteredTemplate;
pub use content_registry::{ContentRepresentations, RegisteredContent};
pub use route::Route;

// This is just a trait alias to help make type signatures a bit saner.
pub trait ByteStream: Stream<Item = Result<Bytes, StreamError>>
where
    Self: Unpin,
{
}
impl<T> ByteStream for T where T: Stream<Item = Result<Bytes, StreamError>> + Unpin {}

/// A piece of rendered content along with its media type.
pub struct Media<Content: ByteStream> {
    pub media_type: MediaType,
    pub content: Content,
}
impl<Content: ByteStream> Media<Content> {
    fn new(media_type: MediaType, content: Content) -> Self {
        Self {
            media_type,
            content,
        }
    }
}

/// Indicates that it was not possible to produce rendered output, either
/// because rendering was attempted and failed or because no acceptable media
/// types are available.
#[derive(Error, Debug)]
pub enum RenderError {
    #[error(transparent)]
    RenderingFailed(RenderingFailedError),

    #[error("The requested content cannot be rendered as an acceptable media type.")]
    CannotProvideAcceptableMediaType,

    #[doc(hidden)]
    #[error("{} This should never happen: {}", bug_message!(), .0)]
    Bug(String),
}

/// Indicates that something went wrong after starting to stream content.
#[derive(Error, Debug)]
pub enum StreamError {
    #[error(
        "Process exited with {}{}",
        match .exit_code {
            Some(code) => format!("code {}", code),
            None => String::from("unknown code"),
        },
        .stderr_contents.as_ref().map(|message| format!(": {}", message)).unwrap_or_default(),
    )]
    ExecutableExitedWithNonzero {
        pid: u32,
        exit_code: Option<i32>,
        stderr_contents: Option<String>,
    },

    #[error("Executable output could not be captured")]
    ExecutableOutputCouldNotBeCaptured { pid: u32 },

    #[error("Input/output error during rendering")]
    IOError {
        #[from]
        source: io::Error,
    },

    #[error("Stream was cancelled")]
    Canceled,
}

pub trait Render {
    type Output;
    fn render<'engine, 'accept, ServerInfo, QueryParameters, Engine, Accept>(
        &self,
        context: RenderContext<'engine, ServerInfo, QueryParameters, Engine>,
        acceptable_media_ranges: Accept,
    ) -> Result<Media<Self::Output>, RenderError>
    where
        ServerInfo: Clone + Serialize,
        QueryParameters: Clone + Serialize,
        Engine: ContentEngine<ServerInfo>,
        Accept: IntoIterator<Item = &'accept MediaRange>,
        Self::Output: ByteStream;
}

// These must match up with serialized property names in RequestData and
// RenderData.
const TARGET_MEDIA_TYPE_PROPERTY_NAME: &str = "target-media-type";
const REQUEST_DATA_PROPERTY_NAME: &str = "request";
const ROUTE_PROPERTY_NAME: &str = "route";
const QUERY_PARAMETERS_PROPERTY_NAME: &str = "query-parameters";

/// Render data that comes from requests.
#[derive(Clone, Serialize)]
#[serde(rename_all = "kebab-case")]
pub struct RequestData<QueryParameters: Clone + Serialize> {
    /// The request [`Route`] that caused this content to be rendered, if any.
    pub route: Option<Route>,

    /// A parsed version of the request URI's query string.
    pub query_parameters: QueryParameters,
}

/// Data passed to handlebars templates and executables.
///
/// Fields serialize into kebab-case (e.g. `server_info` becomes `server-info`).
#[derive(Clone, Serialize)]
#[serde(rename_all = "kebab-case")]
pub struct RenderData<ServerInfo: Clone + Serialize, QueryParameters: Clone + Serialize> {
    /// A hierarchial index of the content. This is serialized with the name
    /// `/` (with handlebars escaping this looks like `[/].[foo/].bar`).
    #[serde(rename = "/")]
    pub index: ContentIndex,

    /// Metadata about the server, such as its version.
    pub server_info: ServerInfo,

    /// The best [`MediaType`] as determined by content negotiation. Rendering
    /// must emit content in this media type.
    pub target_media_type: Option<MediaType>,

    /// Data that comes from requests.
    pub request: RequestData<QueryParameters>,

    /// An [HTTP `4xx` or `5xx` status code](https://datatracker.ietf.org/doc/html/rfc7231#section-6)
    /// indicating that something went wrong. This will be set while rendering
    /// content for the `--error-handler-route`.
    pub error_code: Option<u16>,
}

/// Values used during rendering, including the data passed to handlebars
/// templates and executables.
pub struct RenderContext<'engine, ServerInfo, QueryParameters, Engine>
where
    ServerInfo: Clone + Serialize,
    QueryParameters: Clone + Serialize,
    Engine: ContentEngine<ServerInfo>,
{
    content_engine: &'engine Engine,
    data: RenderData<ServerInfo, QueryParameters>,
}

impl<'engine, ServerInfo, QueryParameters, Engine>
    RenderContext<'engine, ServerInfo, QueryParameters, Engine>
where
    ServerInfo: Clone + Serialize,
    Engine: ContentEngine<ServerInfo>,
    QueryParameters: Clone + Serialize,
{
    pub fn into_error_context(self, error_code: u16) -> Self {
        RenderContext {
            data: RenderData {
                error_code: Some(error_code),
                ..self.data
            },
            ..self
        }
    }
}
