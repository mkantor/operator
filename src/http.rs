use crate::content::*;
use actix_rt::System;
use actix_web::http::header::{self, Header};
use actix_web::{http, web, App, HttpRequest, HttpResponse, HttpServer};
use mime_guess::MimeGuess;
use std::cmp::Ordering;
use std::io::{self, Read};
use std::net::ToSocketAddrs;
use std::sync::{Arc, RwLock};

struct AppData<E: 'static + ContentEngine + Send + Sync> {
    shared_content_engine: Arc<RwLock<E>>,
    index_route: String,
}

pub fn run_server<A, E>(
    shared_content_engine: Arc<RwLock<E>>,
    index_route: &str,
    socket_address: A,
) -> Result<(), io::Error>
where
    A: 'static + ToSocketAddrs,
    E: 'static + ContentEngine + Send + Sync,
{
    let index_route = String::from(index_route);

    log::info!("Initializing HTTP server");
    let mut system = System::new("server");
    let result = system.block_on(async move {
        HttpServer::new(move || {
            App::new()
                .app_data(AppData {
                    shared_content_engine: shared_content_engine.clone(),
                    index_route: index_route.clone(),
                })
                .route("/{path:.*}", web::get().to(get::<E>))
        })
        .bind(socket_address)?
        .run()
        .await
    });

    log::info!("HTTP server has terminated");
    result
}

/// Use the URL path, app data, and accept header to render some content for
/// the HTTP response.
///
/// Content negotiation is performed based on media types (just the accept
/// header; not accept-language, etc) and content is only rendered as media
/// types the client asked for.
///
/// If the path has an extension which maps to a media range it will be
/// considered for content negotiation instead of the accept header. This
/// feature exists because there is not a great way to link to particular
/// representations of resources on the web without putting something in the
/// URL, and it's awfully convenient for humans (compare "to get my resume in
/// PDF format, visit http://mysite.com/resume.pdf" to "...first install this
/// browser extension that lets you customize HTTP headers, then set the accept
/// header to application/pdf, then visit http://mysite.com/resume").
async fn get<E: 'static + ContentEngine + Send + Sync>(request: HttpRequest) -> HttpResponse {
    let app_data = request
        .app_data::<AppData<E>>()
        .expect("App data was not of the expected type!");

    let path = request
        .match_info()
        .get("path")
        .expect("Failed to match request path!");

    let http_version = match request.version() {
        http::Version::HTTP_09 => "HTTP/0.9",
        http::Version::HTTP_10 => "HTTP/1.0",
        http::Version::HTTP_11 => "HTTP/1.1",
        http::Version::HTTP_2 => "HTTP/2.0",
        http::Version::HTTP_3 => "HTTP/3.0",
        _ => "HTTP",
    };
    log::info!(
        "Handling request {} {} {}",
        http_version,
        request.method(),
        request.uri()
    );

    let (route, media_range_from_url) = if path.is_empty() {
        // Default to the index route.
        (app_data.index_route.as_str(), None)
    } else {
        let media_range_from_url = MimeGuess::from_path(path).first();
        let route = if media_range_from_url.is_some() {
            // Drop the extension from the path.
            path.rsplitn(2, '.').last().expect(
                "Calling rsplitn(2, ..) on a non-empty string returned an empty iterator. This should be impossible!"
            )
        } else {
            path
        };
        (route, media_range_from_url)
    };

    // Use the media type from the URL path extension if there was one,
    // otherwise use the accept header.
    let mut parsed_accept_header_value = header::Accept::parse(&request);
    let acceptable_media_ranges = match media_range_from_url {
        Some(ref media_range_from_url) => vec![media_range_from_url],
        None => match parsed_accept_header_value {
            Ok(ref mut accept_value) => acceptable_media_ranges_from_accept_header(accept_value),
            Err(error) => {
                log::warn!(
                    "Malformed accept header value `{:?}` in request for \"/{}\": {}",
                    request.headers().get("accept"),
                    route,
                    error
                );
                return HttpResponse::BadRequest()
                    .content_type(mime::TEXT_PLAIN.essence_str())
                    .body("Malformed accept header value.");
            }
        },
    };

    let render_result = {
        let content_engine = app_data
            .shared_content_engine
            .read()
            .expect("RwLock for ContentEngine has been poisoned");

        content_engine.get(route).map(|content| {
            let render_context = content_engine.get_render_context(route);
            content.render(render_context, acceptable_media_ranges)
        })
    };

    match render_result {
        Some(Ok(Media {
            mut content,
            media_type,
        })) => {
            let mut response_bytes = Vec::new();
            match content.read_to_end(&mut response_bytes) {
                Ok(_) => {
                    log::info!("Successfully rendered /{} as {}", route, media_type);
                    HttpResponse::Ok()
                        .content_type(media_type.to_string())
                        .body(response_bytes)
                }
                Err(error) => {
                    log::error!("Failed to read content for /{}: {}", route, error);
                    HttpResponse::InternalServerError()
                        .content_type("text/plain")
                        .body("Unable to fulfill request.")
                }
            }
        }
        Some(Err(error @ ContentRenderingError::CannotProvideAcceptableMediaType { .. })) => {
            log::warn!("Cannot provide an acceptable response: {}", error);
            HttpResponse::NotAcceptable()
                .content_type("text/plain")
                .body("Cannot provide an acceptable response.")
        }
        Some(Err(error)) => {
            log::warn!("Failed to render content from route /{}: {}", route, error);
            HttpResponse::InternalServerError()
                .content_type("text/plain")
                .body("Unable to fulfill request.")
        }
        None => {
            log::warn!("No content found at /{}", route);
            HttpResponse::NotFound()
                .content_type("text/plain")
                .body("Not found.")
        }
    }
}

fn acceptable_media_ranges_from_accept_header<'a>(
    accept_value: &'a mut header::Accept,
) -> Vec<&'a MediaRange> {
    // If the accept header value is empty, allow any media type.
    if accept_value.is_empty() {
        vec![&mime::STAR_STAR]
    } else {
        // Sort in order of descending quality (so the client's most-preferred
        // representation is first).
        //
        // Note that QualityItem only implements PartialOrd, not Ord. I thought
        // that might be because the parser lossily converts decimal strings into
        // integers (for the `q` parameter), but it turns out the implementation
        // actually never returns None (as of actix-web v3.0.0). If that ever
        // changes and there is some scenario where a pair of items from the
        // accept header can't be ordered then soliton will give them equal
        // preference. ¯\_(ツ)_/¯
        accept_value.sort_by(|a, b| {
            b.partial_cmp(a).unwrap_or_else(|| {
                log::warn!(
                    "Accept header items `{}` and `{}` could not be ordered by quality",
                    a,
                    b
                );
                Ordering::Equal
            })
        });

        accept_value
            .iter()
            .map(|quality_item| &quality_item.item)
            .collect::<Vec<&'a MediaRange>>()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_lib::*;
    use actix_web::body::Body;
    use actix_web::http::StatusCode;
    use actix_web::test::TestRequest;
    use std::path::Path;

    fn test_request(content_directory_path: &Path, url_path: &'static str) -> TestRequest {
        let directory = ContentDirectory::from_root(&content_directory_path).unwrap();
        let shared_content_engine = FilesystemBasedContentEngine::from_content_directory(
            directory,
            SolitonVersion("0.0.0"),
        )
        .expect("Content engine could not be created");

        TestRequest::default()
            .app_data(AppData {
                shared_content_engine: shared_content_engine,
                index_route: String::new(),
            })
            .param("path", url_path)
    }

    #[actix_rt::test]
    async fn content_may_be_not_found() {
        let request =
            test_request(&example_path("empty"), "nothing/exists/at/this/path").to_http_request();
        let response = get::<FilesystemBasedContentEngine>(request).await;

        assert_eq!(response.status(), StatusCode::NOT_FOUND);
    }

    #[actix_rt::test]
    async fn content_can_be_retrieved_with_exact_media_type() {
        let request = test_request(&example_path("hello-world"), "hello")
            .header("accept", "text/html")
            .to_http_request();

        let response = get::<FilesystemBasedContentEngine>(request).await;
        let response_body = response
            .body()
            .as_ref()
            .expect("Response body was not available");
        let response_content_type = response
            .headers()
            .get("content-type")
            .expect("Response was missing content-type header");

        assert_eq!(
            response.status(),
            StatusCode::OK,
            "Response status was not 200 OK"
        );
        assert_eq!(
            response_content_type, "text/html",
            "Response content-type was not text/html",
        );
        assert_eq!(
            response_body,
            &Body::from_slice(b"hello world\n"),
            "Response body was incorrect"
        );
    }

    #[actix_rt::test]
    async fn content_can_be_retrieved_with_media_range() {
        let request = test_request(&example_path("hello-world"), "hello")
            .header("accept", "text/*")
            .to_http_request();

        let response = get::<FilesystemBasedContentEngine>(request).await;
        let response_body = response
            .body()
            .as_ref()
            .expect("Response body was not available");
        let response_content_type = response
            .headers()
            .get("content-type")
            .expect("Response was missing content-type header");

        assert_eq!(
            response.status(),
            StatusCode::OK,
            "Response status was not 200 OK"
        );
        assert_eq!(
            response_content_type, "text/html",
            "Response content-type was not text/html",
        );
        assert_eq!(
            response_body,
            &Body::from_slice(b"hello world\n"),
            "Response body was incorrect"
        );
    }

    #[actix_rt::test]
    async fn content_can_be_retrieved_with_star_star_media_range() {
        let request = test_request(&example_path("hello-world"), "hello")
            .header("accept", "*/*")
            .to_http_request();

        let response = get::<FilesystemBasedContentEngine>(request).await;
        let response_body = response
            .body()
            .as_ref()
            .expect("Response body was not available");
        let response_content_type = response
            .headers()
            .get("content-type")
            .expect("Response was missing content-type header");

        assert_eq!(
            response.status(),
            StatusCode::OK,
            "Response status was not 200 OK"
        );
        assert_eq!(
            response_content_type, "text/html",
            "Response content-type was not text/html",
        );
        assert_eq!(
            response_body,
            &Body::from_slice(b"hello world\n"),
            "Response body was incorrect"
        );
    }

    #[actix_rt::test]
    async fn content_can_be_retrieved_with_elaborate_accept_header() {
        let request = test_request(&example_path("hello-world"), "hello")
            .header("accept", "audio/aac, text/*;q=0.9, image/gif;q=0.1")
            .to_http_request();

        let response = get::<FilesystemBasedContentEngine>(request).await;
        let response_body = response
            .body()
            .as_ref()
            .expect("Response body was not available");
        let response_content_type = response
            .headers()
            .get("content-type")
            .expect("Response was missing content-type header");

        assert_eq!(
            response.status(),
            StatusCode::OK,
            "Response status was not 200 OK"
        );
        assert_eq!(
            response_content_type, "text/html",
            "Response content-type was not text/html",
        );
        assert_eq!(
            response_body,
            &Body::from_slice(b"hello world\n"),
            "Response body was incorrect"
        );
    }

    #[actix_rt::test]
    async fn content_can_be_retrieved_with_missing_accept_header() {
        let request = test_request(&example_path("hello-world"), "hello").to_http_request();

        let response = get::<FilesystemBasedContentEngine>(request).await;
        let response_body = response
            .body()
            .as_ref()
            .expect("Response body was not available");
        let response_content_type = response
            .headers()
            .get("content-type")
            .expect("Response was missing content-type header");

        assert_eq!(
            response.status(),
            StatusCode::OK,
            "Response status was not 200 OK"
        );
        assert_eq!(
            response_content_type, "text/html",
            "Response content-type was not text/html",
        );
        assert_eq!(
            response_body,
            &Body::from_slice(b"hello world\n"),
            "Response body was incorrect"
        );
    }

    #[actix_rt::test]
    async fn multimedia_content_can_be_retrieved() {
        let request = test_request(&example_path("multimedia"), "dramatic-prairie-dog")
            .header("accept", "video/*")
            .to_http_request();

        let response: HttpResponse = get::<FilesystemBasedContentEngine>(request).await;
        let response_body = response
            .body()
            .as_ref()
            .expect("Response body was not available");
        let response_content_type = response
            .headers()
            .get("content-type")
            .expect("Response was missing content-type header");

        assert_eq!(
            response.status(),
            StatusCode::OK,
            "Response status was not 200 OK"
        );
        assert_eq!(
            response_content_type, "video/mp4",
            "Response content-type was not video/mp4",
        );

        let response_bytes = match response_body {
            Body::None => vec![],
            Body::Empty => vec![],
            Body::Bytes(bytes) => bytes.to_vec(),
            Body::Message(_) => {
                unimplemented!("can't get bytes from response with generic message body")
            }
        };
        assert_eq!(
            response_bytes.len(),
            198946,
            "Response body did not have the expected size",
        );
    }

    #[actix_rt::test]
    async fn content_cannot_be_retrieved_if_no_acceptable_media_type() {
        let request = test_request(&example_path("hello-world"), "hello")
            .header("accept", "application/msword, font/otf, audio/3gpp2;q=0.1")
            .to_http_request();

        let response = get::<FilesystemBasedContentEngine>(request).await;

        assert_eq!(
            response.status(),
            StatusCode::NOT_ACCEPTABLE,
            "Response status was not 406 Not Acceptable"
        );
    }

    #[actix_rt::test]
    async fn extension_on_url_takes_precedence_over_accept_header() {
        // Note .html extension on URL path, but no text/html (nor any other
        // workable media range) in the accept header.
        let request = test_request(&example_path("hello-world"), "hello.html")
            .header("accept", "application/msword, font/otf, audio/3gpp2;q=0.1")
            .to_http_request();

        let response = get::<FilesystemBasedContentEngine>(request).await;
        let response_content_type = response
            .headers()
            .get("content-type")
            .expect("Response was missing content-type header");

        assert_eq!(
            response.status(),
            StatusCode::OK,
            "Response status was not 200 OK"
        );
        assert_eq!(
            response_content_type, "text/html",
            "Response content-type was not text/html",
        );
    }

    #[actix_rt::test]
    async fn if_url_has_extension_accept_header_is_ignored() {
        // URL path extension has the wrong media type, but accept header has
        // the correct one. Should be HTTP 406 because the accept header is not
        // considered when there is an extension.
        let request = test_request(&example_path("hello-world"), "hello.doc")
            .header("accept", "text/html")
            .to_http_request();

        let response = get::<FilesystemBasedContentEngine>(request).await;

        assert_eq!(
            response.status(),
            StatusCode::NOT_ACCEPTABLE,
            "Response status was not 406 Not Acceptable"
        );
    }
}
