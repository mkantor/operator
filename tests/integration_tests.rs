mod lib;

use actix_web::client::Client as HttpClient;
use actix_web::http::StatusCode;
use lib::*;
use operator::content::ContentDirectory;
use operator::test_lib::*;
use regex::Regex;
use std::collections::BTreeMap;
use std::env;
use std::io::Write;
use std::process::Stdio;
use std::str;
use test_env_log::test;

#[actix_rt::test]
async fn samples_match_snapshots() {
    for content_directory in sample_content_directories() {
        let content_directory_root = &content_directory.root();

        let log_prefix_regex = {
            let datetime_pattern = r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d+[-+]\d{2}:\d{2}";
            let log_level_pattern = r"ERROR|WARN|INFO|DEBUG|TRACE";
            let log_prefix_pattern =
                format!("^({}) - ({}) - ", datetime_pattern, log_level_pattern);
            Regex::new(&log_prefix_pattern).unwrap()
        };

        let unordered_content = render_everything_for_snapshots(&content_directory).await;
        let contents = unordered_content
            .iter()
            // Files can be omitted from snapshots with a naming convention.
            .filter(|(key, _)| !is_omitted_from_snapshots(key))
            // If dynamic content files mention where the repo is checked
            // out, redact it to keep tests portable.
            .map(|(key, value)| (key, value.replace(PROJECT_DIRECTORY, "$PROJECT_DIRECTORY")))
            // Also redact release modes in paths so we can release with
            // snapshots generated during debug builds.
            .map(|(key, value)| (key, value.replace("target/debug", "target/$PROFILE")))
            .map(|(key, value)| (key, value.replace("target/release", "target/$PROFILE")))
            // Also remove the prefix used on log messages.
            .map(|(key, value)| (key, String::from(log_prefix_regex.replace_all(&value, ""))))
            .collect::<BTreeMap<_, _>>();

        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_input_file(content_directory_root);
        let sample_name = content_directory_root
            .strip_prefix(sample_path("."))
            .expect("Failed to strip samples directory prefix")
            .to_str()
            .expect("Sample path is not UTF-8");
        insta_settings.bind(|| insta::assert_yaml_snapshot!(sample_name, contents));
    }
}

#[test]
fn missing_subcommand_is_error() {
    let mut command = operator_command(&[] as &[&str]);
    let output = command.output().expect("Failed to execute process");

    assert!(
        !output.status.success(),
        "Executing `{:?}` succeeded when it should have failed",
        command
    );
}

#[test]
fn invalid_subcommand_is_error() {
    let mut command = operator_command(&["invalid-subcommand"]);
    let output = command.output().expect("Failed to execute process");

    assert!(
        !output.status.success(),
        "Executing `{:?}` succeeded when it should have failed",
        command
    );
}

#[test]
fn eval_subcommand_succeeds() {
    let input = "{{#if true}}hello world{{/if}}";
    let expected_output = "hello world";

    let mut command = operator_command(&["eval", "--content-directory=/dev/null"]);
    command
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());
    let mut child = command.spawn().expect("Failed to spawn process");
    child
        .stdin
        .as_mut()
        .expect("Failed to capture child process stdin")
        .write_all(input.as_bytes())
        .expect("Failed to write to child process stdin");
    let output = child.wait_with_output().expect("Failed to execute process");

    assert!(
        output.status.success(),
        "Executing `{:?}` failed when it should have succeeded",
        command
    );
    assert_eq!(
        String::from_utf8(output.stdout).expect("Output was not valid UTF-8"),
        expected_output,
        "Executing `{:?}` did not produce the expected output",
        command
    );
}

#[test]
fn get_subcommand_succeeds() {
    let expected_output = "hello world";

    let mut command = operator_command(&[
        "get",
        &format!(
            "--content-directory={}",
            &sample_path("hello-world").to_str().unwrap()
        ),
        "--route=/hello",
    ]);
    let output = command.output().expect("Failed to execute process");

    assert!(
        output.status.success(),
        "Executing `{:?}` failed when it should have succeeded: {}",
        command,
        String::from_utf8(output.stderr).unwrap_or(String::from(
            "Unable to display error message because stderr was not UTF-8"
        ))
    );
    assert_eq!(
        String::from_utf8(output.stdout).expect("Output was not valid UTF-8"),
        expected_output,
        "Executing `{:?}` did not produce the expected output",
        command
    );
}

#[actix_rt::test]
async fn serve_subcommand_succeeds() {
    let content_directory = ContentDirectory::from_root(&sample_path("hello-world")).unwrap();
    let server = RunningServer::start(&content_directory).expect("Server failed to start");

    let expected_response_body = "hello world";

    let request = HttpClient::new()
        .get(format!("http://{}/hello", server.address()))
        .header(
            "Accept",
            "application/msword, text/*;q=0.9, image/gif;q=0.1",
        );

    let mut response = request.send().await.expect("Unable to send HTTP request");
    let response_body = response.body().await.expect("Unable to get response body");
    let response_content_type = response
        .headers()
        .get("Content-Type")
        .expect("Response was missing Content-Type header");

    assert_eq!(
        response.status(),
        StatusCode::OK,
        "Response status was not 200 OK"
    );
    assert_eq!(
        response_content_type, "text/plain",
        "Response Content-Type was not text/plain",
    );
    assert_eq!(
        response_body, expected_response_body,
        "Response body was incorrect"
    );
}
