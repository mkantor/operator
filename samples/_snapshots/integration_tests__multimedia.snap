---
source: tests/integration_tests.rs
expression: contents
input_file: samples/multimedia
---
dramatic-prairie-dog.mp4: binary data with hash 993ffd47041700b
wilhelm-scream.ogg: binary data with hash 9eb0b3ab8b7a0069
